services/gundeck/test/unit/Push.hs

```diff
@@ -0,0 +1,85 @@
+{-# LANGUAGE DataKinds                  #-}
+{-# LANGUAGE FlexibleContexts           #-}
+{-# LANGUAGE GADTs                      #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
+{-# LANGUAGE LambdaCase                 #-}
+{-# LANGUAGE MultiParamTypeClasses      #-}
+{-# LANGUAGE OverloadedStrings          #-}
+{-# LANGUAGE RecordWildCards            #-}
+{-# LANGUAGE ScopedTypeVariables        #-}
+{-# LANGUAGE TemplateHaskell            #-}
+{-# LANGUAGE TupleSections              #-}
+{-# LANGUAGE ViewPatterns               #-}
+
+{-# OPTIONS_GHC -Wno-incomplete-patterns #-}
+
+module Push where
+
+import Imports
+import Gundeck.Push (pushAll, pushAny)
+import Gundeck.Push.Websocket as Web (bulkPush)
+import Gundeck.Types
+import MockGundeck
+import Test.QuickCheck
+import Test.QuickCheck.Instances ()
+import Test.Tasty
+import Test.Tasty.QuickCheck
+
+import qualified Data.Aeson as Aeson
+
+
+tests :: TestTree
+tests = testGroup "bulkpush" $
+    [ testProperty "web sockets" webBulkPushProps
+    , testProperty "native pushes" pushAllProps
+    ]
+
+
+mkEnv :: (Pretty MockEnv -> Property) -> Positive Int -> Property
+mkEnv prop (Positive len) = forAllShrink (Pretty <$> resize len genMockEnv) (shrinkPretty shrinkMockEnv) prop
+
+
+webBulkPushProps :: Positive Int -> Property
+webBulkPushProps plen@(Positive len) = mkEnv mkNotifs plen
+  where
+    mkNotifs :: Pretty MockEnv -> Property
+    mkNotifs (Pretty env) = forAllShrink
+      (Pretty <$> resize len (genNotifs env))
+      (shrinkPretty shrinkNotifs)
+      (webBulkPushProp env)
+
+webBulkPushProp :: MockEnv -> Pretty [(Notification, [Presence])] -> Property
+webBulkPushProp env (Pretty notifs) =
+    counterexample "^ environment, notifications\n" $
```

Comment:
QuickCheck was printing the environment and the list of notifications twice (the first time because it always prints function parameters, the second time because of `counterexample`). I couldn't convince it to stop doing that so I dropped the `counterexample`. Both were formatted the same, luckily, so the tests aren't hurt.